### PR TITLE
ENH : Allow to_sql to recognize single sql type #11886

### PR DESCRIPTION
**Pull request by RahulHP**
Continued in #13252 

---

This solves #11886 
It checks whether the passed dtype variable is a dictionary.
If not, it creates a new dictionary with keys as the columns of the dataframe.
It then passes this dictionary  to the pandasSQL_builder.

### CODE REVIEW

**Comment by jreback** [comment]
tests!

File: pandas/io/sql.py [hunk 1]
```diff
@@ -564,6 +564,10 @@ def to_sql(frame, name, con, flavor='sqlite', schema=None, if_exists='fail',
         raise NotImplementedError("'frame' argument should be either a "
                                   "Series or a DataFrame")
 
+    if dtype and not isinstance(dtype,dict):
+        temp_type = dtype
+        dtype = { col_name : temp_type for col_name in frame }
```
Comment (TomAugspurger) [inline pandas/io/sql.py]: I think flake8 will complain about this line. Need to get rid of the spaces around the curly braces.
Oh, also space after the common above in the `isinstance` check

Comment (MaximilianR) [inline pandas/io/sql.py]: Also `is_dictlike` (which isn't a perfect function but at least improvements will be encapsulated)

**Comment by RahulHP** [comment]
Update: Writing the tests has been more difficult that I thought it would be. I am currently working on it though.

**Comment by jorisvandenbossche** [comment]
@RahulHP For the test, best have

**Comment by RahulHP** [comment]
Update: I added the new test and it passes the Travis build.

File: pandas/io/sql.py [hunk 2]
```diff
@@ -1222,11 +1225,16 @@ def to_sql(self, frame, name, if_exists='fail', index=True,
         chunksize : int, default None
             If not None, then rows will be written in batches of this size at a
             time.  If None, all rows will be written at once.
-        dtype : dict of column name to SQL type, default None
+        dtype : single SQL type or dict of column name to SQL type, default None
             Optional specifying the datatype for columns. The SQL type should
-            be a SQLAlchemy type.
+            be a SQLAlchemy type.If all columns are of the same type, one
```
Comment (jorisvandenbossche) [inline pandas/io/sql.py]: missing space before 'If'

File: pandas/io/sql.py [hunk 3]
```diff
@@ -1222,11 +1225,16 @@ def to_sql(self, frame, name, if_exists='fail', index=True,
...
+        dtype : single SQL type or dict of column name to SQL type, default None
             Optional specifying the datatype for columns. The SQL type should
-            be a SQLAlchemy type.
+            be a SQLAlchemy type.If all columns are of the same type, one 
+            single value can be used.
+
 
         """
+        if dtype and not is_dictlike(dtype):
+            temp_type = dtype
```
Comment (jorisvandenbossche) [inline pandas/io/sql.py]: this `temp_type` is not needed AFAIK, you can just use `dtype` in the line below

File: pandas/io/tests/test_sql.py [hunk 4]
```diff
@@ -1165,6 +1165,21 @@ def test_to_sql(self):
...
+                ('c','d')]
+        df = DataFrame(data,columns=cols)
+        df.to_sql('single_dtype_test',self.conn,dtype=sqlalchemy.TEXT)
+        meta = sqlalchemy.schema.MetaData(bind=self.conn)
+        meta.reflect()
+        sqltypea = meta.tables['single_dtype_test'].columns['A'].type
+        sqltypeb = meta.tables['single_dtype_test'].columns['B'].type
+        self.assertTrue(isinstance(sqltypea, sqlalchemy.TEXT))
+        self.assertTrue(isinstance(sqltypeb, sqlalchemy.TEXT))
+        self.drop_table('single_dtype_test')
```
Comment (jorisvandenbossche) [inline pandas/io/tests/test_sql.py]: Can you put these tests next to the current test for the dtype argument? (to keep things a little bit organised in the test file, it is already difficult to navigate here :-))

**Comment by RahulHP** [comment]
Edits done

**Comment by jreback** [comment]
can you rebase / squash

**Comment by RahulHP** [comment]
Done, I followed the Contributing guide as best as I could but I think I may have messed it up.

**Comment by jorisvandenbossche** [comment]
@RahulHP There indeed slipped in some changes from other PRs. Normally, by doing exactly this, it should be solved:


```
git checkout to_sql-dictionary
git fetch upstream
git rebase upstream/master
git push -f origin to_sql-dictionary
```

**Comment by jreback** [comment]
can you rebase/update

**Comment by RahulHP** [comment]
Sorry, I'll work on this in a few days.

**Comment by jreback** [comment]
pls rebase / update

**Comment by RahulHP** [comment]
I have created a new pull request here https://github.com/pydata/pandas/pull/13252

**Pull request closed without merging**